text/0000-min-target-api-version.md

```diff
@@ -0,0 +1,154 @@
+* Trying to compile code with an implicit dependency on a API version greater than what is supported by the target platform leads to linker errors. For example, the `x86_64-pc-windows-msvc` target's rustc implementation requires `SetThreadErrorMode` which was introduced in Windows 7. This means trying to build the compiler on older versions of Windows will fail with [a less than helpful linker error](https://github.com/rust-lang/rust/issues/35471).
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust targets are often thought of as monoliths. The thought is that if you compile a binary for that target, that binary should be able to run on any system that fits that target's description. However, this is not actually true. For example, when compiling for `x86_64-pc-windows-msvc` and linking with the standard library, my binary has implicitly taken a dependency on a set of APIs that Windows exposes for certainly functionality. If I try to run my binary on older systems that do not have those APIs, then my binary will fail to run. Platforms usually expose a set of APIs in backward compatible with each release. When compiling for a certain target, you are therefore declaring a dependency on a minimum target API version that you rely on for your binary to run. 
+
+By default, the standard library uses a sensible minimum API version. For example, for `x86_64-pc-windows-msvc` the minimum API version is "6.1.7600" which corresponds to Windows 7. However, there's good reason why you might want control of how your code is compiled depending on what the minimum API version is set as. For instance, if you want to:
+
+* set your crate's minimum API version higher than that of the standard library. 
+* change certain implementation details of your crate depending on what a downstream user sets their minimum API version to be.  
+* have some sensible compiler error if users of your crate require a lower minimum API version that you require.
+
+This is where the `cfg` predicate `min_target_api_version` comes in. This allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(min_target_api_version >= "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
```

Comment:
From looking at this code example it's unclear what `min_target_api_version >= "6.0.6000"` refers to. Is it Linux? Windows? Android? Without the comment we have no way to tell.

For the purpose of clarity the platform identifier and the version range should always be co-located. This also makes it more resilient to change over time, as a search for e.g. `windows` will also surface the version range predicate.

I would imagine co-locating the platform and version number could look something like this:

```rust
#[cfg(windows(min = "6.0.6000", max = "7.3.5000))]
```

Reply:
I definitely agree that in this example, this is clearer. I do think there are some questions that would need to be answered though:
* I believe extending the existing `#[cfg(windows]` annotation would be possible in a future compatible way, but this shortened syntax is only available for `windows` and `unix`. 
* For `unix` this doesn't make sense, as there is no unified, versioned thing called `unix`. 
* The cfg predicate this most naturally ties to is `target_os`, but I don't see a particularly natural way to co-locate `min_target_api_version` with `target_os` (outside of existing `any` and `all` constructs).
* This is slightly redundant in code that is already conditionally compiled on `target_os`. For instance, if you have a large chunk of code that is Windows only, and a small piece of that code is Windows 10+ only, then you'll end up repeating the Windows information twice. 

So, I'm sympathetic to this concern, but I think we'll need to think a bit more on the best way to address it.